### DISCUSSION TITLE
Access component variables in component tests

### DISCUSSION THREAD

**Discussion by devjared1** [cypress-io/cypress#24723] [Questions and Help]
We have recently added the Vuetify plugin to our Vue3/Vite app. I have went through the documentation and added the necessary steps to wrap the components in the parent `VApp` component. Here is my `cypress\index.js` file:
```
import { VApp } from 'vuetify/lib/components/VApp/index.mjs';
import { h } from 'vue';
import { mount } from '@cypress/vue';

Cypress.Commands.add("mount", (component, props, args) => {
  return mount(
    { render: () => h(VApp, {}, [h(component, props)]) },
    { ...args }
  );
});
```
TestNavbar.vue file:
```
<template>
  <v-app-bar density="compact" data-cy="navbar">
    <v-app-bar-nav-icon color="grey-lighten-2" data-cy="drawer-btn" />
    <v-app-bar-title class="pl-0">
      <span data-cy="title">Test Title</span>
    </v-app-bar-title>
  </v-app-bar>
</template>

<script setup lang="ts">
import { ref } from 'vue';

const testFoo = ref('');
</script>

<style scoped lang="sass">
</style>
```
TestNavbar.cy.js file
```
import TestNavbar from './TestNavbar.vue';
import vuetify from '../plugins/vuetify';

describe('Navbar component tests', () => {
  it.only('Default view', () => {
    cy.mount(TestNavbar, {}, {
      global: {
        plugins: [vuetify]
      }
    })
      .then(() => {
        cy.get('[data-cy="navbar"]').should('be.visible');
        cy.get('[data-cy="drawer-btn"]').should('be.visible');
        cy.get('[data-cy="title"]').should('be.visible').then((span) => {
          expect(span.text()).to.eq('Test Title');
        });
        // before Vuetify plugin
        console.log(Cypress.vueWrapper.vm.testFoo);
        // after Vuetify plugin, get the nested TestNavbar component
        console.log(Cypress.vueWrapper.getComponent(TestNavbar).vm.testFoo);
      });
  });
});
```
Before we added the Vuetify plugin, I was able to access component variables after mounting with `Cypress.vueWrapper.vm.variableName`. Now, since we have to wrap components with the `VApp` component in tests, `Cypress.vueWrapper.vm` is tied to the `VApp` component. I thought that doing `Cypress.vueWrapper.getComponent(TestNavbar).vm.variableName` would give me access but it is `undefined`.<br><br>
Does anyone have any suggestions on anything else to try? Thanks in advance!

**Comment by portyu9** [comment]
Hi @devjared1 

There are two things happening here.

First, because your custom mount renders:

```js
h(VApp, {}, [h(component, props)])
```

the root Vue Test Utils wrapper represents `VApp`, not `TestNavbar`. Use the wrapper returned by `cy.mount()` and find the child component:

```js
cy.mount(TestNavbar).then(({ wrapper }) => {
  const navbar = wrapper.findComponent(TestNavbar)

  // ...
})
```

Second, your variable is declared in `<script setup>`:

```vue
<script setup lang="ts">
import { ref } from 'vue'

const testFoo = ref('')
</script>
```

Vue components using `<script setup>` are **closed by default**. If you intentionally want that variable available through the public component instance, expose it:

```vue
<script setup lang="ts">
import { ref } from 'vue'

const testFoo = ref('')

defineExpose({
  testFoo,
})
</script>
```

Then:

```js
cy.mount(TestNavbar).then(({ wrapper }) => {
  const navbar = wrapper.findComponent(TestNavbar)

  expect(navbar.vm.testFoo).to.equal('')
})
```

I would also avoid relying on the old global `Cypress.vueWrapper`; current Cypress exposes the Vue Test Utils wrapper directly from the result of `cy.mount()`.

Where possible, prefer asserting the component's visible behavior rather than private refs. `defineExpose()` is appropriate when `testFoo` is intentionally part of the component's public interface.

Please mark this answer as accepted if it helped, thank you.
